File: Shared/Sources/KassaShared/DTOs.swift
import Foundation

// MARK: - Katalog

public struct ArticleDTO: Codable, Hashable, Sendable, Identifiable {
    public var id: String
    public var category: String
    public var name: String
    public var priceCents: Int
    public var sortOrder: Int
    public var active: Bool

    public init(id: String, category: String, name: String, priceCents: Int, sortOrder: Int, active: Bool) {
        self.id = id
        self.category = category
        self.name = name
        self.priceCents = priceCents
        self.sortOrder = sortOrder
        self.active = active
    }

    public var price: Money { Money(cents: priceCents) }
}

// MARK: - Bestellungen

public struct OrderLineDTO: Codable, Hashable, Sendable, Identifiable {
    public var id: UUID
    public var tableNumber: Int
    public var articleId: String
    /// Name und Preis werden bei der Buchung eingefroren: eine spätere
    /// Preisänderung darf alte Rechnungen nicht rückwirkend verändern.
    public var nameSnapshot: String
    public var unitPriceCents: Int
    public var qty: Int
    public var createdAt: Date
    public var deviceId: String
    public var voidedAt: Date?
    public var settlementId: UUID?
    public var updatedSeq: Int

    public init(
        id: UUID,
        tableNumber: Int,
        articleId: String,
        nameSnapshot: String,
        unitPriceCents: Int,
        qty: Int,
        createdAt: Date,
        deviceId: String,
        voidedAt: Date? = nil,
        settlementId: UUID? = nil,
        updatedSeq: Int
    ) {
        self.id = id
        self.tableNumber = tableNumber
        self.articleId = articleId
        self.nameSnapshot = nameSnapshot
        self.unitPriceCents = unitPriceCents
        self.qty = qty
        self.createdAt = createdAt
        self.deviceId = deviceId
        self.voidedAt = voidedAt
        self.settlementId = settlementId
        self.updatedSeq = updatedSeq
    }

    public var lineTotal: Money { Money(cents: unitPriceCents * qty) }
    /// Offen = weder storniert noch kassiert.
    public var isOpen: Bool { voidedAt == nil && settlementId == nil }
}

public struct NewOrderLine: Codable, Hashable, Sendable {
    /// Vom Client vergeben — macht das Anlegen idempotent, auch wenn die
    /// Offline-Queue denselben Command mehrfach schickt.
    public var id: UUID
    public var tableNumber: Int
    public var articleId: String
    public var qty: Int
    public var createdAt: Date

    public init(id: UUID, tableNumber: Int, articleId: String, qty: Int, createdAt: Date) {
        self.id = id
        self.tableNumber = tableNumber
        self.articleId = articleId
        self.qty = qty
        self.createdAt = createdAt
    }
}

public struct CreateOrderLinesRequest: Codable, Sendable {
    public var lines: [NewOrderLine]
    public init(lines: [NewOrderLine]) { self.lines = lines }
}

// MARK: - Kassieren

public struct SettlementLineSelection: Codable, Hashable, Sendable {
    public var lineId: UUID
    /// Teilmenge: kleiner als `qty` der Zeile ist erlaubt, der Server splittet dann.
    public var qty: Int

    public init(lineId: UUID, qty: Int) {
        self.lineId = lineId
        self.qty = qty
    }
}

public struct CreateSettlementRequest: Codable, Sendable {
    public var id: UUID
    public var tableNumber: Int
    public var lines: [SettlementLineSelection]
    /// Vom Client errechnete Summe. Der Server rechnet selbst nach und lehnt
    /// bei Abweichung ab — schützt vor veraltetem Client-Stand.
    public var amountCents: Int
    public var paidAt: Date
    /// Aufrundung des Gastes. Steht bewusst neben `amountCents` und nicht darin,
    /// damit die Nachrechnung am Server weiterhin die reine Zeilensumme prüft.
    public var tipCents: Int

    public init(
        id: UUID,
        tableNumber: Int,
        lines: [SettlementLineSelection],
        amountCents: Int,
        paidAt: Date,
        tipCents: Int = 0
    ) {
        self.id = id
        self.tableNumber = tableNumber
        self.lines = lines
        self.amountCents = amountCents
        self.paidAt = paidAt
        self.tipCents = tipCents
    }

    /// Die Offline-Queue legt diesen Request als rohes `Data` ab. Nach einem
    /// App-Update muss eine alte, noch nicht gesendete Zahlung ohne `tipCents`
    /// weiterhin dekodierbar sein — sonst bleibt sie für immer liegen.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(UUID.self, forKey: .id)
        self.tableNumber = try container.decode(Int.self, forKey: .tableNumber)
        self.lines = try container.decode([SettlementLineSelection].self, forKey: .lines)
        self.amountCents = try container.decode(Int.self, forKey: .amountCents)
        self.paidAt = try container.decode(Date.self, forKey: .paidAt)
        self.tipCents = try container.decodeIfPresent(Int.self, forKey: .tipCents) ?? 0
    }
}

public struct SettlementDTO: Codable, Hashable, Sendable, Identifiable {
    public var id: UUID
    public var tableNumber: Int
    public var totalCents: Int
    public var paidAt: Date
    public var deviceId: String
    /// Betriebstag als `yyyy-MM-dd` — nicht der Kalendertag, siehe Cutoff.
    public var businessDay: String
    public var updatedSeq: Int
    /// Getrennt vom Umsatz, damit `totalCents` weiter gegen die Zeilen aufgeht.
    public var tipCents: Int

    public init(
        id: UUID,
        tableNumber: Int,
        totalCents: Int,
        paidAt: Date,
        deviceId: String,
        businessDay: String,
        updatedSeq: Int,
        tipCents: Int = 0
    ) {
        self.id = id
        self.tableNumber = tableNumber
        self.totalCents = totalCents
        self.paidAt = paidAt
        self.deviceId = deviceId
        self.businessDay = businessDay
        self.updatedSeq = updatedSeq
        self.tipCents = tipCents
    }

    /// Toleranter Decode wie beim Request: ein Client mit älterem Stand darf
    /// eine Antwort ohne `tipCents` nicht als kaputt ansehen.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(UUID.self, forKey: .id)
        self.tableNumber = try container.decode(Int.self, forKey: .tableNumber)
        self.totalCents = try container.decode(Int.self, forKey: .totalCents)
        self.paidAt = try container.decode(Date.self, forKey: .paidAt)
        self.deviceId = try container.decode(String.self, forKey: .deviceId)
        self.businessDay = try container.decode(String.self, forKey: .businessDay)
        self.updatedSeq = try container.decode(Int.self, forKey: .updatedSeq)
        self.tipCents = try container.decodeIfPresent(Int.self, forKey: .tipCents) ?? 0
    }

    public var total: Money { Money(cents: totalCents) }
    public var tip: Money { Money(cents: tipCents) }
    /// Umsatz + Trinkgeld — was tatsächlich in der Kassa liegt.
    public var grandTotal: Money { Money(cents: totalCents + tipCents) }
}

/// Antwortkörper bei HTTP 409 auf `POST /settlements`.
public struct SettlementConflictDTO: Codable, Sendable {
    public var conflictingLineIds: [UUID]
    public var settledByDevice: String?
    public var reason: Reason

    public enum Reason: String, Codable, Sendable {
        case alreadySettled
        case lineVoided
        case amountMismatch
        case unknownLine
        case quantityExceeded
    }

    public init(conflictingLineIds: [UUID], settledByDevice: String?, reason: Reason) {
        self.conflictingLineIds = conflictingLineIds
        self.settledByDevice = settledByDevice
        self.reason = reason
    }
}

// MARK: - Sync

public struct SyncResponse: Codable, Sendable {
    public var lines: [OrderLineDTO]
    public var settlements: [SettlementDTO]
    public var maxSeq: Int
    /// Signalisiert, dass `since` zu alt war und der Client komplett neu laden muss.
    public var isFullReload: Bool

    public init(lines: [OrderLineDTO], settlements: [SettlementDTO], maxSeq: Int, isFullReload: Bool = false) {
        self.lines = lines
        self.settlements = settlements
        self.maxSeq = maxSeq
        self.isFullReload = isFullReload
    }
}

/// Einzige Nachricht über den WebSocket. Er trägt bewusst keinen State, sondern
/// sagt nur "es gibt Neues" — der Client holt sich das Delta dann über /sync.
public struct SyncPing: Codable, Sendable {
    public var seq: Int
    public init(seq: Int) { self.seq = seq }
}

// MARK: - Auth & Konfiguration

public struct LoginRequest: Codable, Sendable {
    public var password: String
    public var deviceName: String
    public init(password: String, deviceName: String) {
        self.password = password
        self.deviceName = deviceName
    }
}

public struct LoginResponse: Codable, Sendable {
    public var token: String
    public var deviceId: String
    public init(token: String, deviceId: String) {
        self.token = token
        self.deviceId = deviceId
    }
}

public struct ServerConfigDTO: Codable, Sendable {
    public var tableCount: Int
    /// Ab welcher Stunde ein neuer Betriebstag beginnt (Default 6).
    public var businessDayCutoffHour: Int
    /// Ändert sich bei jedem CSV-Import — der Client lädt den Katalog dann neu.
    public var catalogVersion: Int

    public init(tableCount: Int, businessDayCutoffHour: Int, catalogVersion: Int) {
        self.tableCount = tableCount
        self.businessDayCutoffHour = businessDayCutoffHour
        self.catalogVersion = catalogVersion
    }
}

/// Geraetename fuer die Anzeige — der Kuechenbon zeigt damit, wer bestellt hat.
/// Bewusst ohne Token-Hash: der verlaesst den Server nie.
public struct DeviceDTO: Codable, Hashable, Sendable, Identifiable {
    public var id: String
    public var name: String

    public init(id: String, name: String) {
        self.id = id
        self.name = name
    }
}

// MARK: - Tagesabschluss

public struct DayReportDTO: Codable, Sendable {
    public var businessDay: String
    /// Reiner Warenumsatz — die Kategoriesummen gehen exakt darauf auf.
    public var totalCents: Int
    public var settlementCount: Int
    public var byCategory: [CategoryTotal]
    public var topArticles: [ArticleTotal]
    public var settlements: [SettlementDTO]
    public var tipCents: Int

    public struct CategoryTotal: Codable, Hashable, Sendable {
        public var category: String
        public var totalCents: Int
        public init(category: String, totalCents: Int) {
            self.category = category
            self.totalCents = totalCents
        }
    }

    public struct ArticleTotal: Codable, Hashable, Sendable {
        public var articleId: String
        public var name: String
        public var qty: Int
        public var totalCents: Int
        public init(articleId: String, name: String, qty: Int, totalCents: Int) {
            self.articleId = articleId
            self.name = name
            self.qty = qty
            self.totalCents = totalCents
        }
    }

    public init(
        businessDay: String,
        totalCents: Int,
        settlementCount: Int,
        byCategory: [CategoryTotal],
        topArticles: [ArticleTotal],
        settlements: [SettlementDTO],
        tipCents: Int = 0
    ) {
        self.businessDay = businessDay
        self.totalCents = totalCents
        self.settlementCount = settlementCount
        self.byCategory = byCategory
        self.topArticles = topArticles
        self.settlements = settlements
        self.tipCents = tipCents
    }

    /// Toleranter Decode: ein Bericht von einem Server ohne Trinkgeld-Feld
    /// bleibt lesbar.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.businessDay = try container.decode(String.self, forKey: .businessDay)
        self.totalCents = try container.decode(Int.self, forKey: .totalCents)
        self.settlementCount = try container.decode(Int.self, forKey: .settlementCount)
        self.byCategory = try container.decode([CategoryTotal].self, forKey: .byCategory)
        self.topArticles = try container.decode([ArticleTotal].self, forKey: .topArticles)
        self.settlements = try container.decode([SettlementDTO].self, forKey: .settlements)
        self.tipCents = try container.decodeIfPresent(Int.self, forKey: .tipCents) ?? 0
    }

    public var total: Money { Money(cents: totalCents) }
    public var tip: Money { Money(cents: tipCents) }
    /// Umsatz + Trinkgeld — was tatsächlich in der Kassa liegt.
    public var grandTotal: Money { Money(cents: totalCents + tipCents) }
}

// MARK: - Fehler

public struct APIErrorDTO: Codable, Sendable {
    public var reason: String
    public init(reason: String) { self.reason = reason }
}

File: Android/app/src/main/java/at/heuriger/kassa/wire/Dtos.kt
@file:UseSerializers(InstantSerializer::class, UuidSerializer::class)

package at.heuriger.kassa.wire

import java.time.Instant
import java.util.UUID
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers

// MARK: - Katalog

@Serializable
data class ArticleDto(
    /** Achtung: String, keine UUID — der Katalog kommt aus einem CSV-Import. */
    val id: String,
    val category: String,
    val name: String,
    val priceCents: Int,
    val sortOrder: Int,
    val active: Boolean,
) {
    val price: Money get() = Money(priceCents)
}

// MARK: - Bestellungen

@Serializable
data class OrderLineDto(
    val id: UUID,
    val tableNumber: Int,
    val articleId: String,
    /**
     * Name und Preis werden bei der Buchung eingefroren: eine spaetere
     * Preisaenderung darf alte Rechnungen nicht rueckwirkend veraendern.
     */
    val nameSnapshot: String,
    val unitPriceCents: Int,
    val qty: Int,
    val createdAt: Instant,
    val deviceId: String,
    val voidedAt: Instant? = null,
    val settlementId: UUID? = null,
    val updatedSeq: Int,
    /**
     * Sonderwunsch des Gastes zu genau dieser Position. Der Server trimmt ihn und
     * kuerzt still auf 120 Zeichen — ein 4xx wuerde die ganze Buchungsrunde
     * dauerhaft in der Offline-Queue parken.
     */
    val note: String? = null,
) {
    val lineTotal: Money get() = Money(unitPriceCents * qty)

    /** Offen = weder storniert noch kassiert. */
    val isOpen: Boolean get() = voidedAt == null && settlementId == null
}

@Serializable
data class NewOrderLine(
    /**
     * Vom Client vergeben — macht das Anlegen idempotent, auch wenn die
     * Offline-Queue denselben Command mehrfach schickt.
     */
    val id: UUID,
    val tableNumber: Int,
    val articleId: String,
    val qty: Int,
    val createdAt: Instant,
    /**
     * Die Notiz haengt an der Bestellzeile, nicht am Bon: den Kuechenbon loest
     * nicht die App aus, sondern der Druckdienst, der `/sync` pollt und selbst
     * entscheidet, welche neuen Zeilen ein Bon werden.
     *
     * Default `null`: die Offline-Queue haelt alte Payloads als rohes JSON, die
     * kennen `note` noch nicht und muessen weiter lesbar bleiben.
     */
    val note: String? = null,
)

@Serializable
data class CreateOrderLinesRequest(
    val lines: List<NewOrderLine>,
)

// MARK: - Kassieren

@Serializable
data class SettlementLineSelection(
    val lineId: UUID,
    /** Teilmenge: kleiner als `qty` der Zeile ist erlaubt, der Server splittet dann. */
    val qty: Int,
)

@Serializable
data class CreateSettlementRequest(
    val id: UUID,
    val tableNumber: Int,
    val lines: List<SettlementLineSelection>,
    /**
     * Vom Client errechnete Summe. Der Server rechnet selbst nach und lehnt
     * bei Abweichung ab — schuetzt vor veraltetem Client-Stand.
     */
    val amountCents: Int,
    val paidAt: Instant,
    /**
     * Aufrundung des Gastes. Steht bewusst neben `amountCents` und nicht darin,
     * damit die Nachrechnung am Server weiterhin die reine Zeilensumme prueft.
     *
     * Default 0: die Offline-Queue haelt alte Payloads als rohes JSON, die
     * kennen `tipCents` noch nicht und muessen weiter lesbar bleiben.
     */
    val tipCents: Int = 0,
)

@Serializable
data class SettlementDto(
    val id: UUID,
    val tableNumber: Int,
    val totalCents: Int,
    val paidAt: Instant,
    val deviceId: String,
    /** Betriebstag als `yyyy-MM-dd` — nicht der Kalendertag, siehe Cutoff. */
    val businessDay: String,
    val updatedSeq: Int,
    /** Getrennt vom Umsatz, damit `totalCents` weiter gegen die Zeilen aufgeht. */
    val tipCents: Int = 0,
) {
    val total: Money get() = Money(totalCents)
    val tip: Money get() = Money(tipCents)

    /** Umsatz + Trinkgeld — was tatsaechlich in der Kassa liegt. */
    val grandTotal: Money get() = Money(totalCents + tipCents)
}

/** Antwortkoerper bei HTTP 409 auf `POST /settlements`. */
@Serializable
data class SettlementConflictDto(
    val conflictingLineIds: List<UUID>,
    val settledByDevice: String? = null,
    val reason: Reason,
) {
    @Serializable
    enum class Reason {
        @SerialName("alreadySettled") ALREADY_SETTLED,
        @SerialName("lineVoided") LINE_VOIDED,
        @SerialName("amountMismatch") AMOUNT_MISMATCH,
        @SerialName("unknownLine") UNKNOWN_LINE,
        @SerialName("quantityExceeded") QUANTITY_EXCEEDED,
    }
}

// MARK: - Druckauftrag

@Serializable
data class CreatePrintRequestRequest(
    /**
     * Vom Client vergeben — macht das Anlegen idempotent, auch wenn der Kellner
     * den Knopf zweimal drueckt oder die Offline-Queue nachliefert.
     */
    val id: UUID,
    val tableNumber: Int,
    /**
     * Dieselbe Auswahl wie beim Kassieren — die Aufstellung zeigt genau das,
     * was gleich bezahlt wuerde.
     */
    val lines: List<SettlementLineSelection>,
    val requestedAt: Instant,
)

/**
 * Was der Druckdienst bekommt. Ausdruecklich kein Beleg, sondern nur eine
 * Aufstellung zum Nachrechnen — deshalb ohne Betriebstag und Steuerfelder.
 */
@Serializable
data class PrintRequestDto(
    val id: UUID,
    val tableNumber: Int,
    val items: List<Item>,
    val totalCents: Int,
    val requestedAt: Instant,
    val deviceId: String,
    val updatedSeq: Int,
) {
    /**
     * Name und Preis loest der Server aus den Zeilen auf und friert sie hier ein,
     * wie bei `OrderLineDto.nameSnapshot`: der Ausdruck muss zeigen, was gebucht
     * wurde, nicht was der Katalog heute sagt.
     */
    @Serializable
    data class Item(
        val name: String,
        val qty: Int,
        val unitPriceCents: Int,
    ) {
        val lineTotal: Money get() = Money(unitPriceCents * qty)
    }

    val total: Money get() = Money(totalCents)
}

// MARK: - Druckauftrag Tagesstatistik

@Serializable
data class CreateDayReportPrintRequestRequest(
    /**
     * Vom Client vergeben, idempotent wie ueberall — ein zweimal gedrueckter
     * Knopf darf keinen zweiten Zettel erzeugen.
     */
    val id: UUID,
    /**
     * "yyyy-MM-dd", der im DatePicker gewaehlte Betriebstag. Gedruckt wird, was
     * am Bildschirm steht, nicht zwingend der heutige Tag.
     */
    val businessDay: String,
    val requestedAt: Instant,
)

/**
 * Traegt bewusst keine Zahlen. Der Bericht wird nicht eingefroren, sondern vom
 * Druckdienst ueber `GET /reports/day` frisch geholt — ein gespeicherter
 * Umsatz-Schnappschuss waere eine zweite Wahrheit neben der berechneten und
 * kaeme einem Journal naeher, als dieses System je sein will.
 */
@Serializable
data class DayReportPrintRequestDto(
    val id: UUID,
    val businessDay: String,
    val requestedAt: Instant,
    val deviceId: String,
    val updatedSeq: Int,
)

// MARK: - Sync

@Serializable
data class SyncResponse(
    val lines: List<OrderLineDto>,
    val settlements: List<SettlementDto>,
    val maxSeq: Int,
    /** Signalisiert, dass `since` zu alt war und der Client komplett neu laden muss. */
    val isFullReload: Boolean = false,
)

/**
 * Einzige Nachricht ueber den WebSocket. Er traegt bewusst keinen State, sondern
 * sagt nur "es gibt Neues" — der Client holt sich das Delta dann ueber /sync.
 */
@Serializable
data class SyncPing(
    val seq: Int,
)

// MARK: - Auth & Konfiguration

@Serializable
data class LoginRequest(
    val password: String,
    val deviceName: String,
)

@Serializable
data class LoginResponse(
    val token: String,
    val deviceId: String,
)

@Serializable
data class ServerConfigDto(
    val tableCount: Int,
    /** Ab welcher Stunde ein neuer Betriebstag beginnt (Default 6). */
    val businessDayCutoffHour: Int,
    /** Aendert sich bei jedem CSV-Import — der Client laedt den Katalog dann neu. */
    val catalogVersion: Int,
)

/**
 * Geraetename fuer die Anzeige — der Kuechenbon zeigt damit, wer bestellt hat.
 * Bewusst ohne Token-Hash: der verlaesst den Server nie.
 */
@Serializable
data class DeviceDto(
    /** Achtung: String, keine UUID. */
    val id: String,
    val name: String,
)

// MARK: - Tagesabschluss

@Serializable
data class DayReportDto(
    val businessDay: String,
    /** Reiner Warenumsatz — die Kategoriesummen gehen exakt darauf auf. */
    val totalCents: Int,
    val settlementCount: Int,
    val byCategory: List<CategoryTotal>,
    val topArticles: List<ArticleTotal>,
    val settlements: List<SettlementDto>,
    /** Default 0: ein Bericht von einem Server ohne Trinkgeld-Feld bleibt lesbar. */
    val tipCents: Int = 0,
) {
    @Serializable
    data class CategoryTotal(
        val category: String,
        val totalCents: Int,
    )

    @Serializable
    data class ArticleTotal(
        val articleId: String,
        val name: String,
        val qty: Int,
        val totalCents: Int,
    )

    val total: Money get() = Money(totalCents)
    val tip: Money get() = Money(tipCents)

    /** Umsatz + Trinkgeld — was tatsaechlich in der Kassa liegt. */
    val grandTotal: Money get() = Money(totalCents + tipCents)
}

// MARK: - Fehler

@Serializable
data class ApiErrorDto(
    val reason: String,
)
